snuba/replacer.py

```diff
@@ -16,13 +21,138 @@
 
 logger = logging.getLogger("snuba.replacer")
 
+executor = ThreadPoolExecutor()
+
+
+class InsertExecutor(ABC):
+    """
+    Executes the Replacement insert query.
+
+    Each implementation provides a different execution policy.
+    """
+
+    @abstractmethod
+    def execute(self, replacement: Replacement, record_counts: int) -> int:
+        """
+        Executes the query according to the policy implemented by this
+        class.
+        """
+        raise NotImplementedError
+
+
+# Used by InsertExecutors to run the query
+RunQuery = Callable[[ClickhousePool, str, int, MetricsBackend], None]
+
+
+class DistributedTableExecutor(InsertExecutor):
+    """
+    InsertExecutor that runs one query only on a distributed table.
+    This relies on Clickhouse to forward the replacement query to
+    each storage node.
+    """
+
+    def __init__(
+        self,
+        runner: RunQuery,
+        connection: ClickhousePool,
+        table: str,
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__connection = connection
+        self.__table = table
+        self.__metrics = metrics
+        self.__runner = runner
+
+    def execute(self, replacement: Replacement, records_count: int) -> int:
+        query = replacement.get_insert_query(self.__table)
+        if query is None:
+            return 0
+        self.__runner(self.__connection, query, records_count, self.__metrics)
+        return records_count
```

Comment:
Why do we return records_count here? It's not modified anywhere in this flow.

Reply:
Because this method can also return 0 (line 69).

The count of of actually modified rows is used here 
https://github.com/getsentry/snuba/pull/2025/files#diff-efbe37619db8bd83053dfc7e8eecf1ec37700b2d5078eb8ad6a5e462ac93605aR284

And if the `get_insert_query` method returns None, then count has to be set to 0.